Adds the option to encode enums as 1-key maps.

ogoodman opened the pull request:
Since enums are part of the Rust language but not part of the CBOR or JSON spec, each serializer has to make a choice how to encode them. 

serde_cbor chose `["Variant", items..]` while serde_json chose `{"Variant": item}` or `{"Variant": [items..]}`. Although this is immaterial when passing data between Rust programs, it becomes visible if `Values` are used, or when data is passed between Rust and other languages.

This turned out to be a problem for me when I tried to migrate a communicating pair of Python and Rust programs from JSON to CBOR, as assumptions about the encoding of enums had already become buried throughout the Python code.

This pull request adds the option (default false) to encode enums in the serde_json way and adds support for automatically decoding enums from either encoding.

pyfisch commented:
Sorry. I missed this. I will review this week.

pyfisch commented:
I think if we add this the serialization should match the one used by serde-json.

ogoodman commented on src/ser.rs:
```diff
@@ -51,6 +51,27 @@ where
     value.serialize(&mut ser)
 }
 
+/// Serializes a value without names to a writer.
```
I've removed the `to_writer_*` functions that were part of this PR.

ogoodman commented on src/ser.rs:
```diff
@@ -96,10 +117,42 @@ where
+    /// Struct fields and enum variants are identified by their numeric indices rather than names
+    /// to save space.
+    pub packed: bool,
+    /// When enum_as_map is set, enums are encoded as maps rather than arrays.
```
Done.

ogoodman commented:
Tests added. I've also added a `to_vec` function to `SerializerOptions` and implemented `Default` on the latter. That makes it possible for people to code in such a way that adding new options won't be a breaking change (although that requires that they add `..Default::default()`). The alternative is having private fields and a bunch of builder methods, but I don't love that either.

kostko commented:
@ogoodman thanks for taking the time to update this, I just got back from vacation and didn't have time to push my changes before. I did a cursory look over the new changes and it seems ok.

Regarding option defaults, either the `Default` way or the builder seem fine to me with a slight preference for `Default` due to requiring less code, especially for these simple flags that are currently implemented.

ogoodman commented:
@kostko thanks for your comments. Sorry if I caused you to waste time working on this. I didn't think I'd have time to work on it for a while, but then I unexpectedly did.

pyfisch commented:
Thank you @ogoodman :+1:
